sidecar/codec.go

```diff
@@ -15,77 +16,107 @@ const (
 func EncodeToString(t *Ticket) (string, error) {
-	var buf bytes.Buffer
-
-	// Add a prefix to make the ticket easily recognizable by human eyes.
-	if _, err := buf.Write(sidecarPrefix); err != nil {
+	var (
+		checksumBuf bytes.Buffer
+		encodeBuf   bytes.Buffer
+	)
+
+	// First, we'll write the sidecar prefix, as well as the serialized
+	// ticket into the buffer that we'll use to generate the checksum. We
+	// do this as we won't encode the checksum using base58.
+	if _, err := checksumBuf.Write([]byte(sidecarPrefix)); err != nil {
 		return "", err
 	}
-
-	// Serialize the raw ticket itself.
-	if err := SerializeTicket(&buf, t); err != nil {
+	if _, err := checksumBuf.Write(encodingVersion); err != nil {
+		return "", err
+	}
+	if err := SerializeTicket(&checksumBuf, t); err != nil {
 		return "", err
 	}
 
-	// Create a checksum (SHA256) of all the bytes so far (prefix + raw
-	// ticket) and add 4 bytes of that to the string encoded version. This
-	// will prevent the ticket from appearing valid even if parts of it were
-	// not copy/pasted correctly.
-	hash := sha256.New()
-	_, _ = hash.Write(buf.Bytes())
-	checksum := hash.Sum(nil)[0:4]
-	if _, err := buf.Write(checksum); err != nil {
+	// We'll now hash our full checksum buffer to generate the checksum
+	// that we'll tack onto the end of the ticket.
+	checksum := sha256.Sum256(checksumBuf.Bytes())
+
+	// Now that we have the checksum, we'll write the raw ticket and the
+	// checksum to the buffer that we'll use for encoding.
+	//
+	// Similar to Bitcoin, we'll add a version so the prefix of the string
+	// is always the same.
+	if _, err := encodeBuf.Write(encodingVersion); err != nil {
+		return "", err
+	}
+	if err := SerializeTicket(&encodeBuf, t); err != nil {
+		return "", err
+	}
+	if _, err := encodeBuf.Write(checksum[:checksumLen]); err != nil {
 		return "", err
 	}
 
-	return defaultStringEncoding.EncodeToString(buf.Bytes()), nil
+	// The final ticket is the sidecar encoding (non base58 encoded) with
+	// the base58 encoded ticket following it.
+	baseTicket := base58.Encode(encodeBuf.Bytes())
+	return sidecarPrefix + baseTicket, nil
 }
 
 // DecodeString decodes and then deserializes the given sidecar ticket string.
 func DecodeString(s string) (*Ticket, error) {
-	rawBytes, err := defaultStringEncoding.DecodeString(s)
-	if err != nil {
-		return nil, err
+	prefixLength := len(sidecarPrefix)
+	if len(s) < prefixLength {
+		return nil, fmt.Errorf("string contains invalid prefix")
 	}
 
-	if len(rawBytes) < len(sidecarPrefix)+checksumLen {
-		return nil, fmt.Errorf("not a sidecar ticket, invalid length")
+	// First, we'll decode the set of raw bytes from the base58 encoding,
+	// snipping off our custom prefix before decoding.
+	rawBytes := base58.Decode(s[prefixLength:])
+
+	// We'll also ensure that the length passes basic sanity checks to not
+	// trip up the slicing logic below.
+	expectedLen := len(sidecarPrefix) + len(encodingVersion) + checksumLen
```

Comment:
The sidecar prefix isn't encoded in the payload so it shouldn't be part of this length?

Reply:
Nice catch, removed. Added a quickcheck test as well to make sure we don't crash w/ arbitrary string input.